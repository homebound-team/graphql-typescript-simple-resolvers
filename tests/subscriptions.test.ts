import { createTestSchema, runPlugin } from "./utils/test-helpers";

describe("GraphQL Subscriptions", () => {
  const schemaWithSubscriptions = `
    type User {
      id: ID!
      name: String!
      email: String!
    }

    type Post {
      id: ID!
      title: String!
      author: User!
      content: String!
    }

    type Comment {
      id: ID!
      post: Post!
      author: User!
      text: String!
    }

    union ActivityItem = Post | Comment

    type Query {
      users: [User!]!
      posts: [Post!]!
    }

    type Mutation {
      createUser(name: String!, email: String!): User!
      createPost(title: String!, content: String!, authorId: ID!): Post!
      createComment(postId: ID!, text: String!, authorId: ID!): Comment!
    }

    type Subscription {
      # Simple subscription without arguments
      userCreated: User!

      # Subscription with arguments
      postCreated(authorId: ID): Post!

      # Subscription returning arrays
      newActivity: [ActivityItem!]!

      # Subscription with multiple arguments
      commentsForPost(postId: ID!, authorId: ID): [Comment!]!

      # Subscription returning nullable types
      latestPost: Post
    }
  `;

  it("generates subscription types", async () => {
    const schema = createTestSchema(schemaWithSubscriptions);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface Subscription {
  userCreated: User;
  postCreated: Post;
  newActivity: ActivityItem[];
  commentsForPost: Comment[];
  latestPost: Post | null | undefined;
}`);
  });

  it("generates subscription resolvers without arguments", async () => {
    const schema = createTestSchema(schemaWithSubscriptions);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`userCreated: SubscriptionResolver<Subscription, {}, User>;`);
  });

  it("generates subscription resolvers with arguments", async () => {
    const schema = createTestSchema(schemaWithSubscriptions);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`postCreated: SubscriptionResolver<Subscription, SubscriptionPostCreatedArgs, Post>;`);
  });

  it("generates subscription resolvers with multiple arguments", async () => {
    const schema = createTestSchema(schemaWithSubscriptions);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(
      `commentsForPost: SubscriptionResolver<Subscription, SubscriptionCommentsForPostArgs, readonly Comment[]>;`,
    );
  });

  it("generates argument types for subscriptions with parameters", async () => {
    const schema = createTestSchema(schemaWithSubscriptions);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface SubscriptionPostCreatedArgs {
  authorId?: string | null | undefined;
}`);

    expect(code).toContain(`export interface SubscriptionCommentsForPostArgs {
  postId: string;
  authorId?: string | null | undefined;
}`);
  });

  it("snapshots subscription generation", async () => {
    const schema = createTestSchema(schemaWithSubscriptions);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });
    expect(code).toMatchSnapshot();
  });

  it("handles subscription with mapped types", async () => {
    const schema = createTestSchema(schemaWithSubscriptions);
    const code = await runPlugin(schema, {
      mappers: {
        User: "./entities#UserEntity",
        Post: "./entities#PostEntity",
      },
      scalars: {},
      enumValues: {},
    });

    expect(code).toContain(`import { PostEntity, UserEntity } from "./entities";`);
  });
});
