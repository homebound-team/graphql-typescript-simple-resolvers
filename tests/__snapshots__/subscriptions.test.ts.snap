// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`GraphQL Subscriptions snapshots subscription generation 1`] = `
"import { GraphQLResolveInfo } from "graphql";
import { Context } from "./context";

export interface Resolvers {
  Query: QueryResolvers;
  Mutation: MutationResolvers;
  User?: UserResolvers;
  Post?: PostResolvers;
  Comment?: CommentResolvers;
  Subscription?: SubscriptionResolvers;
}

export type UnionResolvers = { ActivityItem: { __resolveType(o: Post | Comment): string } };

export interface QueryResolvers {
  users: Resolver<{}, {}, readonly User[]>;
  posts: Resolver<{}, {}, readonly Post[]>;
}

export interface MutationResolvers {
  createUser: Resolver<{}, MutationCreateUserArgs, User>;
  createPost: Resolver<{}, MutationCreatePostArgs, Post>;
  createComment: Resolver<{}, MutationCreateCommentArgs, Comment>;
}

export interface UserResolvers {
  id: Resolver<User, {}, string>;
  name: Resolver<User, {}, string>;
  email: Resolver<User, {}, string>;
}

export interface PostResolvers {
  id: Resolver<Post, {}, string>;
  title: Resolver<Post, {}, string>;
  author: Resolver<Post, {}, User>;
  content: Resolver<Post, {}, string>;
}

export interface CommentResolvers {
  id: Resolver<Comment, {}, string>;
  post: Resolver<Comment, {}, Post>;
  author: Resolver<Comment, {}, User>;
  text: Resolver<Comment, {}, string>;
}

export interface SubscriptionResolvers {
  userCreated: SubscriptionResolver<Subscription, {}, User>;
  postCreated: SubscriptionResolver<Subscription, SubscriptionPostCreatedArgs, Post>;
  newActivity: SubscriptionResolver<Subscription, {}, readonly ActivityItem[]>;
  commentsForPost: SubscriptionResolver<Subscription, SubscriptionCommentsForPostArgs, readonly Comment[]>;
  latestPost: SubscriptionResolver<Subscription, {}, Post | null | undefined>;
}

type MaybePromise<T> = T | Promise<T>;
export type Resolver<R, A, T> = (root: R, args: A, ctx: Context, info: GraphQLResolveInfo) => MaybePromise<T>;

export type SubscriptionResolverFilter<R, A, T> = (
  root: R | undefined,
  args: A,
  ctx: Context,
  info: GraphQLResolveInfo,
) => boolean | Promise<boolean>;
export type SubscriptionResolver<R, A, T> = {
  subscribe: (root: R | undefined, args: A, ctx: Context, info: GraphQLResolveInfo) => AsyncIterator<T>;
};

export interface MutationCreateUserArgs {
  name: string;
  email: string;
}
export interface MutationCreatePostArgs {
  title: string;
  content: string;
  authorId: string;
}
export interface MutationCreateCommentArgs {
  postId: string;
  text: string;
  authorId: string;
}
export interface SubscriptionPostCreatedArgs {
  authorId?: string | null | undefined;
}
export interface SubscriptionCommentsForPostArgs {
  postId: string;
  authorId?: string | null | undefined;
}
export interface User {
  id: string;
  name: string;
  email: string;
}

export interface Post {
  id: string;
  title: string;
  author: User;
  content: string;
}

export interface Comment {
  id: string;
  post: Post;
  author: User;
  text: string;
}

export interface Subscription {
  userCreated: User;
  postCreated: Post;
  newActivity: ActivityItem[];
  commentsForPost: Comment[];
  latestPost: Post | null | undefined;
}

export type ActivityItem = Post | Comment;
"
`;
